yeatmanlab/roar-dashboard
# Add Firebase App Check

## Proposed changes
Companion PR: ~~https://github.com/yeatmanlab/roar-firekit/pull/135~~ ~~https://github.com/yeatmanlab/roar-firekit/pull/146~~

Docs PR: https://github.com/yeatmanlab/roar-docs/pull/15

**MAJOR CHANGES**:

- This PR sets up Firebase App Check for use within Firekite and the Dashboard. App Check uses reCAPTCHA tokens to verify that all requests are coming from a list of allowed domains.
- The list of pre-approved domains is `roar.education`, `roar-staging.web.app `for production App Check tokens.
- `localhost` is allowed when developing locally using the debug token, which devs will need in their local `.env` file.
- For PR deploy urls, a new job within the `Build and Deploy` workflow dynamically adds the PR link to the list of approved domains, ensuring that any duplicates are removed. This is achieved with the `gcloud cli` using Firebase credentials stored as repo secrets.


**HOW TO TEST ON PROD**:
1. Open PR link
3. Login as any account type
4. Open the dev console and inspect the network tab for an outgoing request to `https://firestore.googleapis.com/**/*`
5. Check that the headers of the request includes a field for `X-Firebase-AppCheck:` with a token; every request should have this token.
6. Monitor requests here: https://console.firebase.google.com/u/0/project/gse-roar-admin/appcheck/products, https://console.firebase.google.com/u/0/project/gse-roar-assessment/appcheck/products
7. Once ENFORCED mode is on, all requests must have this app check token or they will fail.

**HOW TO TEST ON DEV**:
1. Checkout PR code
2. `npm run dev`
3. Open developer console
4. If you have the debug token saved to your local `.env` file, you should see a console.log() including the debug token. You are successfully bypassing app check in development.
5. If you do not have the debug token saved to your local `.env` file, all requests should fail (?)



## Types of changes

What types of changes does this pull request introduce?




## Checklist




## Justification of missing checklist items



## Further comments
